app/src/org/commcare/adapters/EntityKeyFilterer.java

```diff
@@ -0,0 +1,73 @@
+package org.commcare.adapters;
+
+import android.app.Activity;
+
+import org.commcare.models.Entity;
+import org.commcare.models.NodeEntityFactory;
+import org.javarosa.core.model.instance.TreeReference;
+
+import java.util.ArrayList;
+import java.util.Hashtable;
+import java.util.LinkedHashSet;
+import java.util.List;
+
+/**
+ * Filter (and order) entities by those whose 'extra key', most likely case id,
+ * is present in the provided ordered key set
+ *
+ * @author Phillip Mates (pmates@dimagi.com).
+ */
+public class EntityKeyFilterer extends EntityFiltererBase {
+    private final LinkedHashSet<String> orderedKeySet;
+    private final List<Entity<TreeReference>> matchList = new ArrayList<>();
```

Comment:
I think this can/should be a protected field in `EntityFiltererBase` instead of declaring it in both of the subclasses